## 14 неделя "Численное интегрирование", дополнительный материал для чтения

1. Решение дифференциальных уравнений
2. Сеточные методы ещё более высокой точности 
3. Метод Ромберга 
4. Ограниченная применимость сеточных методов: многомерный случай
5. Другие способы численного расчёта интегралов 


### 1. Решение дифференциальных уравнений

Заметим, что численное интегрирование открывает возможность решения дифференциальных уравнений с начальным условием:

$$
\frac{dy(x)}{dx} = f(x),
\\
y(0) = y_0
\\
y(x) = y_0 + \int \limits_0^x f(x') dx'.
$$

Воспользуемся определением производной и снова возьмём конечное, но достаточно малое значение шага:

$$
\frac{dy(x)}{dx} = \lim_{\Delta x \to 0} \frac{y(x+\Delta x) - y(x)}{\Delta x},
\\
f(x) = \frac{dy(x)}{dx} \simeq \frac{y(x+\Delta x) - y(x)}{\Delta x},
\\
y(x+\Delta x) \simeq y(x) + \Delta x \cdot f(x)
\\
y(x) = y_0 + \Delta x \cdot \sum \limits_{i=0}^{n=x/\Delta x} f(x')
$$

Это - так называемый <b>метод Эйлера</b>, простейший метод решения дифференциальных уравнений с начальным условием
(более правильное название - "явный метод Эйлера").

Фактически это метод прямоугольников интегрирования на сетке! Так же, как и метод прямоугольников, метод Эйлера имеет невысокую точность, но в определённых случаях его достаточно для получения первых данных о явлении или грубой оценки величины. При должной аккуратности его вполне можно рекомендовать к применению, хотя надо понимать, что он обладает существенными недостатками.

<i> <b>В следующем семестре</b> мы вернёмся к этой теме и будем успешно решать достаточно сложные дифференциальные уравнения для хороших современных физических задач.</i> 

<b>Ключевые слова для дальнейшего чтения:</b> <i> конечные разности, методы Эйлера и Рунге-Кутты</i> (см. Ильина, Силаев, том 2)


### 2. Сеточные методы ещё более высокой точности

Если увеличивать количество слагаемых ряда Тейлора подынтегральной функции и улучшать точность приближения далее, то мы получим ещё два известных метода: так называемый "метод Симпсона 3/8" (4-го порядка точности) и метод Буля (5-го порядка).

Добавим их в таблицу, приведённую в основном материале:

|Метод| Число точек| &nbsp; &nbsp; &nbsp; &nbsp;  Коэффициенты &nbsp; &nbsp; &nbsp; &nbsp;   | Точность|
|:-|:-:|:--:|---|
|Прямоугольников|2|h|$\sim h^1$ |
|Трапеций|2| $ \frac{h}{2}\begin{bmatrix}1 & 1 \end{bmatrix} $|$\sim h^2$|
|Симпсона|3|$ \frac{h}{3}\begin{bmatrix}1 & 4 & 1 \end{bmatrix} $| $\sim h^3$|
|Симпсона "3/8"|4|$ \frac{3h}{8}\begin{bmatrix}1 & 3 & 3 & 1 \end{bmatrix} $| $\sim h^4$|
|Буля|5|$ \frac{2h}{45}\begin{bmatrix}7 & 32 & 12 & 32 & 7 \end{bmatrix} $| $\sim h^5$|


Как видно, интегрирование на равномерной сетке имеет варианты с высокой точностью расчёта. Однако сложные схемы с нетривиальными и к тому же меняющимися коэффициентами не очень удобны в реализации и это может привести к ошибкам. Поэтому рекомендуют использовать другой подход:

### 3. Метод Ромберга

Заметим, что поправка к ответу у решёточных методов зависит от величины шага $h$ степенным образом:
$$
\int \limits_a^b f(x) dx 
= h \sum \limits_i f(x_i) + O(h^n)
$$
Тогда мы можем найти эту степенную поправку при помощи приближения полиномом, воспользовавшись навыками с прошлого занятия!

Рассчитаем наш интеграл методом трапеций несколько раз с разными величинами шага $h$.
Полученные значения будут равны
$$
J(h) = h \sum \limits_i f(x_i) = \int \limits_a^b f(x) dx - O(h^2)
$$

Далее приблизим рассчитанные значения $J(h)$ полиномом второй степени (например, при помощи функции polyfit):

$$
J(h) \simeq J(0) + ah + b h^2
$$

Положив в этой формуле $h=0$, получим точное значение интеграла!

Лучше использовать метод Ромберга, чем Симпсона и остальных более высокого порядка, т.к. он состоит из двух простых частей без возможности ошибиться.


### 4. Ограниченная применимость сеточных методов: многомерный случай

Обратите внимание, что методы интегрирования на сетке не очень хорошо годятся для расчёта многомерных интегралов.
Действительно, если в одномерном случае для получения интеграла с точностью 6..8 знаков достаточно использовать разбиение на $n\sim 10^4$ точек, то применение такого же подхода на трёх осях потребует рассчитывать и суммировать уже $n^3\sim 10^{12}$ значений функции, что заметно труднее. Для дальнейшего повышения точности число слагаемых придётся увеличивать непропорционально (вспомним задачу БДЗ с прошлого занятия об оценках трудоёмкости алгоритма).

Более того, часто бывают нужны интегралы большей кратности (например, во многих задачах квантовой физики нужно интегрировать по 6-мерному пространству, а в задачах теоретической физики кратность интегралов вообще не ограничена). 

В таком случае используются другие способы интегрирования, из которых прежде всего стоит упомянуть два:
* разнообразные аналитические преобразования, в том числе разложение подынтегральной функции в ряд,
* и метод Монте-Карло, где для обхода всей интегрируемой области используется генератор случайных чисел.

<i><b>В следующем семестре</b> мы изучим эти темы более подробно.</i>




### 5. Другие способы численного расчёта интегралов

Даже в одномерном случае не все интегралы хорошо рассчитываются при помощи сеточных методов: например, медленно сходятся или имеют другие особенности.
Но те, которые нужны в человеческой деятельности и часто возникают в физических и инженерных расчётах, уже проанализированы математиками и для них были получены альтернативные представления, которые приведены в справочниках и рекомендуются к использованию.
Т.е., конечно же, на интегрировании математика не заканчивается: имеются представления в виде рядов, цепных дробей, рекуррентных соотношений и т.д.

<b>(5а)</b> Так, например, т.н. <b>функция ошибок</b>, возникающая в математической статистике и теоретической физике:
$$
erf(x) = \frac{2}{\sqrt{\pi}} \int \limits_0^x { e^{-t^2} dt }
$$
не очень хорошо рассчитывается сеточными методами (попробуйте и убедитесь сами!).

Для расчёта функции ошибок существуют другие подходы. Так, в справочнике [Wolfram MathWorld](https://mathworld.wolfram.com/Erf.html) даны несколько полезных формул:

- Для малых значений аргумента - ряд (Внимание! на момент 08.08.2022 формула на сайте содержит опечатку - вместо двойного факториала должен быть обычный, в чём легко убедиться, сравнивая две части формулы)
$$
erf(x \ll 1) \simeq
\frac{1}{2\sqrt{\pi}}
e^{-x^2}
\sum \limits_{n=0}^{\infty} \frac{(2x)^{2n+1}}{(2n+1)!}
=
\frac{1}{\sqrt{\pi}}
e^{-x^2}
\left[
    x + \frac{2x^3}{1\cdot 3} + \frac{4 x^5}{1\cdot3\cdot5} + \dots
\right].
$$

- Для любых значений аргумента - цепная дробь:
$$
erf(x) \simeq
1 - \frac{e^{-x^2}}{\sqrt{\pi}}
\cdot
\frac{1}{
    x + 
    \frac{2}{
        2x + 
        \frac{3}{
            x + 
            \frac{4}{
                2x + \dots
            }
        }
    }
}
$$


##### <b>(5b)</b> Другая важная функция, так называемый <b>"интегральный синус"</b>:
$$
Si(x) = \int \limits_0^x \frac{\sin(t)}{t} dt
$$
Этот интеграл часто возникает в оптике и лазерной физике.
Основные свойства: [Wolfram MathWorld](https://mathworld.wolfram.com/SineIntegral.html)

Если нужны значения $Si(x)$ с высокой точностью, то расчёт сеточным методом тоже оказывается недостаточно хорош.
Впрочем, недавно появился новый рецепт! См. статью 2014 года в банке препринтов arxiv.org: https://arxiv.org/pdf/1407.7676.pdf - на странице 36 приведены две громоздкие формулы, дающие значения интегрального синуса с точностью 15..16 значащих цифр на всей числовой оси (одна для $x<4$, вторая для $x>4$). Эти формулы были получены авторами препринта с помощью компьютерного пакета символьных вычислений Maple. 

Ниже приводится функция, реализующая эти формулы:

In [ ]:
import numpy as np

def Si(x):
    if (x<=4):
        x2 = x*x
        numerator = 1 + x2 * (-4.54393409816329991e-2
            + x2 * (1.15457225751016682e-3 + x2 * (-1.41018536821330254e-5
            + x2 * (9.43280809438713025e-8 + x2 * (-3.53201978997168357e-10
            + x2 * (7.08240282274875911e-13 + x2 * (-6.05338212010422477e-16
            )))))))
        denominator = 1 + x2 * (1.01162145739225565e-2 
            + x2 * (4.99175116169755106e-5 + x2 * (1.55654986308745614e-7
            + x2 * (3.28067571055789734e-10 + x2 * (4.5049097575386581e-13
            + x2 * (3.21107051193712168e-16 ))))))
        
        return x*numerator/denominator
    else:
        xm2 = 1/(x*x)
        numerator = 1 + xm2 * (7.44437068161936700618e2  
            + xm2 * (1.96396372895146869801e5 + xm2 * (2.37750310125431834034e7
            + xm2 * (1.43073403821274636888e9 + xm2 * (4.33736238870432522765e10 
            + xm2 * (6.40533830574022022911e11 + xm2 * (4.20968180571076940208e12 
            + xm2 * (1.00795182980368574617e13 + xm2 * (4.94816688199951963482e12 
            - xm2 * 4.94701168645415959931e11 )))))))))

        denominator = 1 + xm2 * (7.46437068161927678031e2  
            + xm2 * (1.97865247031583951450e5 + xm2 * (2.41535670165126845144e7  
            + xm2 * (1.47478952192985464958e9 + xm2 * (4.58595115847765779830e10 
            + xm2 * (7.08501308149515401563e11 + xm2 * (5.06084464593475076774e12 
            + xm2 * (1.43468549171581016479e13 + xm2 * 1.11535493509914254097e13 ))))))))

        f = numerator /( x * denominator )

        numerator = 1 + xm2 * (8.1359520115168615e2   
            + xm2 * (2.35239181626478200e5 + xm2 * (3.12557570795778731e7  
            + xm2 * (2.06297595146763354e9  + xm2 * (6.83052205423625007e10 
            + xm2 * (1.09049528450362786e12 + xm2 * (7.57664583257834349e12 
            + xm2 * (1.81004487464664575e13 + xm2 * (6.43291613143049485e12 
            - xm2 * 1.36517137670871689e12 )))))))))

        denominator = 1 + xm2 * (8.19595201151451564e2  
            + xm2 * (2.40036752835578777e5 + xm2 * (3.26026661647090822e7  
            + xm2 * (2.23355543278099360e9 + xm2 * (7.87465017341829930e10 
            + xm2 * (1.39866710696414565e12 + xm2 * (1.17164723371736605e13 
            + xm2 * (4.01839087307656620e13 + xm2 *  3.99653257887490811e13 ))))))))

        g = xm2 * numerator / denominator
        
        return np.pi/2 - f*np.cos(x) - g*np.sin(x)

print('Si(0.4)=', Si(0.4), ' Si(1)=', Si(1))
print('Si(10)=', Si(10), ' Si(100)=', Si(100))

Si(0.4)= 0.3964614647513729  Si(1)= 0.946083070367183
Si(10)= 1.658347594218874  Si(100)= 1.5622254668890563


Сравним со значениями, рассчитываемыми wolframapha.com:

*    [Si(0.4)](https://www.wolframalpha.com/input?i=Si%280.4%29)=0.3964614647513729...,
    
*    [Si(1)](https://www.wolframalpha.com/input?i=Si%281%29)=0.946083070367183...,
    
*    [Si(10)](https://www.wolframalpha.com/input?i=Si%2810%29)=1.658347594218874...,
    
*    [Si(100)](https://www.wolframalpha.com/input?i=Si%28100%29)=1.5622254668890563...

Как видно, числа полностью совпадают, формула действительно даёт 15-16 значащих цифр!

